### PR TITLE
#3 - lazy symmetric interval hull

### LINKED ISSUE #3: Symbolic symmetric interval hull
The symmetric interval hull is computed explicitly in [Approximations.box_approximation_helper](https://github.com/mforets/LazySets.jl/blob/master/src/Approximations.jl#L226), through 2n support function evaluations.

An application of a new lazy (total or partial) type, symmetric interval hull, is to speedup the discretization phase; the explicit form can take up to 97% of the time for n>10000.

### PR DESCRIPTION
**Pull request by schillic**
Closes #3.

### CODE REVIEW

**Comment by schillic** [comment]
Reminder: This type needs to be adapted to #122.

**Comment by mforets** [comment]
let's rebase when #122 gets merged

**Comment by schillic** [comment]
Done, but this has not advanced and is not ready.

**Comment by schillic** [comment]
Was quite easy in the end 😀
@mforets: Can you have a look?

**Comment by schillic** [comment]
The breaking build is due to outakes in Travis CI. We can restart when they are resolved.

**Comment by mforets** [comment]
i think the auxiliary type should be a single one, maybe a `Dict`, that has tuple values `(index, radius)`? the type and main loop become a bit more compact and you avoid having to choose `radius = [0,0..0]` for a freshly initialized set

**Comment by mforets** [comment]
should we use a cache for `length(d)` in `σ(d::AbstractVector{N}, sih::SymmetricIntervalHull)`, similarly to `zero_N` and `one_N`? (BTW i don't know if `d[i] == zero_N` the same as `d[i] == 0`)

**Comment by schillic** [comment]
A `Dict` does not have constant-time access, and it only hides from you that it also allocates and initializes memory.
Would a `Vector{Tuple}` or `Tuple{Vector, Vector}` be ok?
Let me add that this is an implementation detail. The user does not have to work with these variables (and if Julia had the concept of private variables, these would definitely be private).

**Comment by schillic** [comment]
I assume that most vector implementations don't actually compute anything for `length`, so the compiler might inline this. But ok, we can store it, does not cost much 👍

**Comment by schillic** [comment]
I rebased with master and squashed my old commits. Now I will add the subtyping from `Hyperrectangle` that you proposed.

File: src/SymmetricIntervalHull.jl [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+### Algorithm
+
+We ask for the support vector of the underlying set for both the positive and
+negative unit vector in the dimension `i`.
+"""
+function compute_radius(sih::SymmetricIntervalHull,
+                        i::Int,
+                        dimension::Int=dim(sih))
+    one_N = one(N)
```
Comment (mforets) [inline src/SymmetricIntervalHull.jl]: the type `N` will not be known there. we can pass it?

Comment (schillic) [reply src/SymmetricIntervalHull.jl]: Yes, I will fix this.

File: src/SymmetricIntervalHull.jl [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    right_bound = σ(sparsevec([i], [one_N], dimension), sih.X)
+    left_bound = σ(sparsevec([i], [-one_N], dimension), sih.X)
+    sih.radius[i] = max(right_bound[i], abs(left_bound[i]))
+    sih.known[i] = true
```
Comment (mforets) [inline src/SymmetricIntervalHull.jl]: should we invert these last two lines? such that it returns the computed radius which is more interesting.

Comment (schillic) [reply src/SymmetricIntervalHull.jl]: Actually, the docs said that it returns the radius...

**Comment by schillic** [comment]
This one is ready again, but I did not change the auxiliary fields yet. What should we do here?

**Pull request merged**